cuqi/solver/_solver.py

```diff
@@ -320,6 +327,136 @@ def solve(self):
         return x, k
 
 
+class PCGLS:
+    """Conjugate Gradient method for least squares problems with preconditioning
+    
+    See Bjorck (1996) - Numerical Methods for Least Squares Problems. Pag 294.
+
+    Parameters
+    ----------
+    A : ndarray or callable f(x,*args).
+        Function or array representing the forward model.
+    b : ndarray
+        Data vector.
+    x0 : ndarray
+        Initial guess.    
+    P : ndarray
+        Preconditioner array in sparse format.
+    maxit : int
+        The maximum number of iterations.
+    tol : float
+        The numerical tolerance for convergence checks.
+    """    
+    def __init__(self, A, b, x0, P, maxit, tol=1e-6, shift=0):
+        self.A = A
+        self.b = b
+        self.x0 = x0
+        self.P = P
+        self.maxit = int(maxit)
+        self.tol = tol        
+        self.shift = shift
+        self.dim = len(x0)
+        if not callable(A):
+            self.explicitA = True
+        else:
+            self.explicitA = False
+        #
+        if self.dim < config.MAX_DIM_INV:
+            self.explicitPinv = True
+            Pinv = spa.linalg.inv(P)
+        else:
+            self.explicitPinv = False
+            Pinv = None # we do cholesky.solve or sparse.solve with P
+            if has_cholmod:
+                # turn P into a cholesky object
+                P = cholesky(P, ordering_method='natural')
+        self.Pinv = Pinv # inverse of the preconditioner as a ndarray or function
```

Comment:
Normally, if there is dependency in the attributes (for example Pinv depends on P) we create setters and getters for them so that the depndenant property can be inferred correctly if the original property changed. Or we can also  make those properties private to indicate that the user is not supposed to change them after initialization. For the scenario when the user create the solver: `mysolver = PCGLS(...)` then afterwards they do something like: `my solver.A = ...`,  `self.explicitPinv` needs to be updated which can be done through a setter. However, I suggest for now we make all such properties private (e.g.: `self._A`, `self._x0`, `self._P`,`self._Pinv`, `self._explicitPinv`,  `self._explicitA`, `self._dim`). And we can provide getters if needed. 

for example for A, in the initializer:
 self._A = A

and if needed we add a property
```python
@property
def A(self):
    return self._A
```
Here the user cannot set A but can retrieve A after initialization.

Reply:
Hi @amal-ghamdi, thanks for the comment. As you suggested, I made the properties private for the moment since it is the easiest solution so I can focus on the horseshoe. I am not really familiar with coding properties, setters and getters, so I guess this has to be implemented for all classes in solver.py at some future re-work.